numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -194,8 +194,18 @@ call_array_function(PyObject* argument, PyObject* method,
         return array_function_method_impl(public_api, types, args, kwargs);
     }
     else {
-        return PyObject_CallFunctionObjArgs(
+        PyObject *errmsg = PyObject_CallFunctionObjArgs(
             method, argument, public_api, types, args, kwargs, NULL);
+        PyObject *err = PyErr_Occurred();
+        if (err && PyErr_GivenExceptionMatches(err, PyExc_TypeError)) {
```

Comment:
This would match every `TypeError`, which could be very slow, and is hacky. I'd suggest something along the lines of `if getattr(type, __array_function_like__, False)`.

Reply:
Of course, this would require an update to the NEP.

Reply:
That would also work, the downside to such an approach is that we would need to then follow an all-or-nothing approach, so if the downstream library implements the `__array_function_like__` attribute, then all creation functions would need to add `like=`. That's not necessarily bad, but it increases effort for adoptability and forces us to maintain a list of all functions that add `like=` in NEP-35, so maintainers know which functions need to add that argument.

I would like to hear more opinions and potentially some other drawbacks from folks before we decide on what path to take.

Reply:
Not sure I like idea of an additional flag much. I guess this is a bit special due to the assumption that you do not want to force projects to add `like=` if it is just ignored anyway.

Right now, adding `kwargs.pop("like")` in your `__array_function__` implementation is also possible (using the fact that `like` isn't used for anything else at the moment).  I guess there is a transition problem possible where projects do generalize to these functions, it just never happened yet. And forcing to add `like=` means they have to catch up. But, in a sense I guess the whole setup is prone to small transition issues like that already.

Reply:
To be clear, you'd only force projects to add `like=` to functions _iff_ they also choose to add `__array_function_like__` attribute. For example, CuPy wouldn't implement either one, but Dask would probably add `__array_function_like__`, and only then it would be forced to add `like=` to all the functions that NEP-35 covers. IOW, it would be opt-in only and once the library decides to add `__array_function_like__`, they would know already of the associated cost.

Reply:
Yeah, the question is whether its worth avoiding the inconvenience for most `__array_function__` implementers by baking in one of these proposals.  I still feel it may not be right now.  If you do not want to keep a list of such functions, we could also tag on `__array_function_uses_like_argument__` to the functions to signal it (but again, that would move the need to ignore the kwarg to the project).

Reply:
@seberg is correct in that, we don't want that projects _need_ to implement that (e.g., CuPy), but having `like=` being passed downstream can be useful for other projects (e.g., Dask).

Reply:
Let me be a bit more concrete about my concerns here: It may silence other `TypeError`s, even for Dask. Suppose that CuPy is a bit peculiar and raises a `TypeError` where NumPy would not. Dask dispatches to CuPy because of the `like=` kwarg. CuPy raises a `TypeError`. The same thing gets dispatched without the `like=` kwarg and succeeds, thus silencing the original error, and silently not doing what the user would like.

I would really prefer a user (by which I mean duck-array-side) controllable switch as to whether or not to do `kwargs.pop('like')`.

Reply:
@seberg:

> to manually drop the kwarg in __array_function__.

This is what I'm suggesting the warning message would advocate doing.

Reply:
@hameerabbasi I agree, this is definitely not a great solution. It's annoying that there isn't an exclusive exception in Python to identify that. If this was simply on the Python side, we could just parse the error message, but I'm not sure if this is something we can/should do in C.

Reply:
> This is what I'm suggesting the warning message would advocate doing.

If we are to do a non-backwards compatible change, maybe a better alternative would be to remove `like=` by default, except when the downstream library has some attribute then. This change would reduce any maintenance issues with existing libraries that don't need `like=` (e.g., CuPy and probably the majority of other libraries that implement `__array_function__` today) and just push requirements to those that would benefit from `like=` being passed (possibly only Dask?). The idea of checking for an `__array_function_like__` attribute isn't bad, but it would require `like=` to be added to the signature of all functions where NumPy does too, maybe it would be better if we had a per-function attribute, then downstream can implement that progressively, maybe something like `__array_like__`, `__asarray_like__`, and so on. This would probably mean we need to keep a list of functions that support `like=` in NEP-35, but it's probably a smaller price to pay and I would be fine with maintaining that myself.

Reply:
I have to admit, I still fail a bit to see that it is helpful enough to downstream to be worth the additional complexity.  And I guess Eric has that opinion as well, noting that we can add a deprecation warning (using the proposed logic) to transition more smoothly?

At this time, all you need is add `kwargs.pop("like")` (that is, after all what NumPy currently does).  I suppose there is some chance that dropping the `like` argument should be a per-function decision in the future, but even that does not seem a huge hassle?

I do admit, it is one more thing a downstream implementer would have to check off the TODO list when implementing `__array_function__` and we will have to document that...

Reply:
I understand that point, and to be fair I'm not necessarily against either of the solutions. Going for the warning raises a few different points though:

1. We're now effectively breaking backwards compatibility (which was something I was trying hard to avoid), specifically with NEP-18 in this case, we would need to state that in NEP-35 and perhaps even amend NEP-18;
1. The warning will be _very_ annoying for users as all calls to NumPy array creation will emit one, probably thousands in any workflow.

I think the complexity in NumPy can probably be kept to around a dozen lines and that would let us avoid breaking backwards compatibility and avoid the warning, which feels like a much lower maintenance cost than updating NEP-18 to me. But again, I'm not sold to the idea of not breaking backwards compatibility either, I'm happy to hear more arguments in favor/against both if we think we need a bit more time to make the right decision.

Reply:
@pentschev I am surprised it should be so super annoying to users?  This is a new feature, there may have been implicit support for this before, but basically it can only show up if:

1. The array-like already supported NEP-18 even though it was not implemented (which is simple enough, so often true)
2. The user opts to use `like=`.  Or `dask` opts in, and e.g. its dependency `cupy` only has implicit support.

DeprecationWarnings should not be visible to most end-users, however, since that is the default and in `Dask` + `cupy` scenario it is only necessary to update `cupy`?

I am currently unconcerned about backward compatibility, since NEP 35 was never released and the `DeprecationWarning` gives at least a bit of a transitional compatibility, but maybe I am underestimating the size of the backcompat break.

Reply:
Maybe I'm overly concerned about things, I'm not sure yet. I believe you're right on the situations where this may show up, I guess I'm also considering the case where CuPy is older than NumPy 1.20, for example, if one were to upgrade to NumPy master now and remain with CuPy in release version, when using a Dask function that does some implicit `like=`-related calls to CuPy, this would be a hassle.

It would only be necessary to do that change in CuPy, Dask would be fine as long as it does add `like=` to all functions where NumPy does too, or remove `like=` on a per-function basis.

As for backward compatibility, what worries me is that for a reader of NEP-18 there was nothing as `like=`, but now we're introducing a change that requires an amendment stating they need to explicitly remove `like=`, I'm not sure if that's ideal, maybe we can rely solely on the `DeprecationWarning` should we follow that path?

Reply:
Well, NEP-18 explicitly does not include `np.array`, if you support it you are beyond what it covers? Also I expect NEP-18 says that signatures should match, which might suggest you have to add the `like=` kwarg as well.

Reply:
@pentschev I talked about this with @seberg at today's triage meeting, we decided that this PR isn't needed. Any function with `like=` implies that if `like=` is passed in, it's the only thing being dispatched on. Therefore, inside `__array_function__`, `self` is equivalent to the `like` that was passed in. So anytime we need `like=`, we could use `self` instead, and copy chunk-types from that. Does that work for your use-case?

Reply:
@hameerabbasi wow, that's indeed a great idea! I'm now embarrassed I didn't think of it before, but thankfully we have you here to come up with that! I agree, this seems to entirely cover that situation, we can then do one of two things in the `__array_function__` definition of those libraries (e.g., Dask) that may eventually have an use for it:

1. Inspect the function to check whether it implements the `like=` argument and do `func(*args, **kwargs, like=self)`; or
1. Special case functions that will use it it and pass `func(*args, **kwargs, like=self)`, in case we it's not possible to do introspection or that's too detrimental for performance.

I'll try to do a bit more testing tomorrow just to ensure I'm not overlooking anything. After that, I'll then close this PR and I'm thinking of making a change to NEP-35 only to add the information on how downstream libraries can achieve this result that we're trying to do for Dask, should someone else need this in the future.

Reply:
Sorry it took me longer than I expected to come back to this. I did some more checking, and I think @hameerabbasi 's solution indeed covers the cases I had in mind. With that said, I just opened https://github.com/numpy/numpy/pull/17725 to add instructions for downstream library maintainers on how to do that, and I'm closing this PR. Thanks @hameerabbasi @seberg and @eric-wieser for feedback here!